### PR TITLE
Reusable workflow build llvm libraries

### PR DESCRIPTION
**Pull request by TianlongLiang**
It depends on PR #1543 . 
As part of issue #1510

### CODE REVIEW

File: .github/workflows/_build_llvm_libraries.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      - name: Build llvm and clang from source
+        id: build_llvm
+        if: ${{ steps.cache_llvm.outputs.cache-hit != 'true' }} && ${{ !inputs.build-lldb }}
+        run: /usr/bin/env python3 ./build_llvm.py --arch X86 WebAssembly
+        working-directory: build-scripts
+
+      - name: Build llvm and clang from source, and build lldb too
+        id: build_llvm_plus_lldb
+        if: ${{ steps.cache_llvm.outputs.cache-hit != 'true' }} && ${{ inputs.build-lldb }}
+        run: /usr/bin/env python3 ./build_llvm.py --arch X86 WebAssembly --project clang lldb
```
Comment (lum1n0us) [inline .github/workflows/_build_llvm_libraries.yml:53]: there will be a specific workflow for lldb building. So we can simplify this line to
`/usr/bin/env python3 ./build_llvm.py --arch X86`

File: .github/workflows/_build_llvm_libraries.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+on:
+  workflow_call:
+    inputs:
+      runs-on:
+        required: true
+        # can't pass array now, workaround by passing Json string and deserialize later
+        default: "['ubuntu-20.04', 'ubuntu-22.04']"
+        type: string
+      build-lldb:
```
Comment (lum1n0us) [inline .github/workflows/_build_llvm_libraries.yml:16]: there will be a specific workflow for lldb building. So `build-lldb` could be removed.

File: .github/workflows/_build_llvm_libraries.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        id: cache_llvm
+        uses: actions/cache@v3
+        with:
+          path: |
+            ./core/deps/llvm/build/bin
+            ./core/deps/llvm/build/include
+            ./core/deps/llvm/build/lib
+            ./core/deps/llvm/build/libexec
+            ./core/deps/llvm/build/share
+          key: ${{ matrix.os }}-${{ env.LLVM_CACHE_SUFFIX }}
```
Comment (lum1n0us) [inline .github/workflows/_build_llvm_libraries.yml:42]: `env.LLVM_CACHE_SUFFIX` is an implicit dependencies and we should only pass dependencies via `inputs`

File: .github/workflows/_build_llvm_libraries.yml (hunk 1 above)
Comment (TianlongLiang) [reply .github/workflows/_build_llvm_libraries.yml:53]: so should I add lldb workflow for this PR or remove all lldb-related modifications in this PR?

Comment (lum1n0us) [reply .github/workflows/_build_llvm_libraries.yml:53]: please  remove all lldb-related modifications in this PR?

File: .github/workflows/_build_llvm_libraries.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# CALLED WORKFLOW
```
Comment (wenyongh) [inline .github/workflows/_build_llvm_libraries.yml:1]: Should add license header like other yml files.

Rename `_build_llvm_libraries.yml` to `build_llvm_libraries.yml`

File: .github/workflows/build_llvm_libraries.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Copyright (C) 2019 Intel Corporation.  All rights reserved.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+name: Reusable workflow-build_llvm_libraries&build_lldb_if_needed
```
Comment (lum1n0us) [inline .github/workflows/build_llvm_libraries.yml:3]:
```suggestion
name: Reusable workflow-build_llvm_libraries
```

File: .github/workflows/build_llvm_libraries.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+on:
+  workflow_call:
+    inputs:
+      runs-on:
+        required: true
+        # can't pass array now, workaround by passing Json string and deserialize later
+        default: "['ubuntu-20.04', 'ubuntu-22.04']"
```
Comment (lum1n0us) [inline .github/workflows/build_llvm_libraries.yml:11]: if `required` is `true`, `default` values will never be used. may consider to remove `default`.

File: .github/workflows/build_llvm_libraries.yml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            ./core/deps/llvm/build/bin
+            ./core/deps/llvm/build/include
+            ./core/deps/llvm/build/lib
+            ./core/deps/llvm/build/libexec
+            ./core/deps/llvm/build/share
+          key: ${{ matrix.os }}-${{ inputs.LLVM_CACHE_SUFFIX }}
+
+      - name: Build llvm and clang from source
+        id: build_llvm
+        if: ${{ steps.cache_llvm.outputs.cache-hit != 'true' }} && ${{ !inputs.build-lldb }}
```
Comment (lum1n0us) [inline .github/workflows/build_llvm_libraries.yml:42]:
```suggestion
        if: ${{ steps.cache_llvm.outputs.cache-hit != 'true' }}
```

File: .github/workflows/compilation_on_android_ubuntu.yml [hunk 8]
```diff
@@ -59,61 +59,17 @@ jobs:
...
-        run: |
-          if [[ ${{ github.repository }} == */wasm-micro-runtime ]]; then
-            echo "::set-output name=light::green"
-          else
-            echo "::set-output name=light::green"
-          fi
-
   build_llvm_libraries:
-    needs: check_repo
     runs-on: ${{ matrix.os }}
```
Comment (lum1n0us) [inline .github/workflows/compilation_on_android_ubuntu.yml:63]:
```suggestion
    uses: ./.github/workflows/build_llvm_libraries.yml
    # ... ...
```

File: .github/workflows/compilation_on_sgx.yml [hunk 9]
```diff
@@ -51,47 +51,17 @@ jobs:
...
-        run: |
-          if [[ ${{ github.repository }} == */wasm-micro-runtime ]]; then
-            echo "::set-output name=light::green"
-          else
-            echo "::set-output name=light::green"
-          fi
-
   build_llvm_libraries:
-    needs: check_repo
     runs-on: ${{ matrix.os }}
```
Comment (lum1n0us) [inline .github/workflows/compilation_on_sgx.yml:55]:
```suggestion
    uses: ./.github/workflows/build_llvm_libraries.yml
    # ... ...
```

**Review by lum1n0us — CHANGES_REQUESTED** [review]
Please rebase the PR to the latest commit of [bytecodealliance:dev/binary_release](https://github.com/bytecodealliance/wasm-micro-runtime/tree/dev/binary_release)

**Comment by wenyongh** [comment]
LGTM

**Pull request merged**